src/modules/compliance/src/lib/Evaluator.cpp

```diff
@@ -16,241 +18,563 @@
     if ((nullptr == name) || (nullptr == value))
     {
-        OsConfigLogError(log, "Rule name or value is null");
+        OsConfigLogError(mContext.GetLogHandle(), "Rule name or value is null");
         return Error("Rule name or value is null");
     }
 
-    if (!strcmp(name, "anyOf"))
+    if (!strcmp(name, "anyOf") || !strcmp(name, "allOf"))
     {
-        if (json_value_get_type(value) != JSONArray)
+        mIndicators.Push(name);
+        const auto result = EvaluateList(value, action, !strcmp(name, "anyOf") ? ListAction::AnyOf : ListAction::AllOf);
+        if (!result.HasValue())
         {
-            logstream << "ERROR: anyOf value is not an array";
-            OsConfigLogError(log, "anyOf value is not an array");
-            return Error("anyOf value is not an array");
+            OsConfigLogError(mContext.GetLogHandle(), "Evaluation failed: %s", result.Error().message.c_str());
+            return result;
         }
-        JSON_Array* array = json_value_get_array(value);
-        size_t count = json_array_get_count(array);
-        logstream << "{ anyOf: [";
-        for (size_t i = 0; i < count; ++i)
+        mIndicators.Back().status = result.Value();
+        mIndicators.Pop();
+        return result.Value();
+    }
+
+    if (!strcmp(name, "not"))
+    {
+        mIndicators.Push("not");
+        const auto result = EvaluateNot(value, action);
+        if (!result.HasValue())
         {
-            JSON_Object* subObject = json_array_get_object(array, i);
-            auto result = EvaluateProcedure(subObject, action);
-            if (!result.HasValue())
-            {
-                logstream << "] == FAILURE }";
-                return result;
-            }
-            if (result.Value() == Status::NonCompliant)
-            {
-                if (i < count - 1)
-                {
-                    logstream << ", ";
-                }
-            }
-            else
-            {
-                logstream << "] == TRUE }";
-                return Status::Compliant;
-            }
+            OsConfigLogError(mContext.GetLogHandle(), "Evaluation failed: %s", result.Error().message.c_str());
+            return result;
         }
-        logstream << "] == FALSE }";
+        mIndicators.Back().status = result.Value();
+        mIndicators.Pop();
+        return result.Value();
+    }
+
+    mIndicators.Push(name);
+    auto result = EvaluateBuiltinProcedure(name, value, action);
+    if (!result.HasValue())
+    {
+        OsConfigLogError(mContext.GetLogHandle(), "Evaluation failed: %s", result.Error().message.c_str());
+        return result;
+    }
+    mIndicators.Back().status = result.Value();
+    mIndicators.Pop();
+
+    return result.Value();
+}
+
+Result<Status> Evaluator::EvaluateList(const json_value_t* value, const Action action, const ListAction listAction)
+{
+    const char* actionName = listAction == ListAction::AnyOf ? "anyOf" : "allOf";
```

Comment:
>listAction == ListAction::AnyOf [](http://example.com/codeflow?start=29&length=31)

nit: could increase code clarity a bit adding a pair of () round braces around this.